allennlp/common/push_to_hf.py

```diff
@@ -0,0 +1,145 @@
+"""
+Utilities for pushing models to the Hugging Face Hub ([hf.co](https://hf.co/)).
+"""
+
+import logging
+import sys
+from typing import Optional, Union
+from pathlib import Path
+
+from allennlp.common.file_utils import cached_path
+import shutil
+
+import zipfile
+import tarfile
+import tempfile
+
+from huggingface_hub import Repository, HfApi, HfFolder
+
+logger = logging.getLogger(__name__)
+
+README_TEMPLATE = """---
+tags:
+- allennlp
+---
+
+# TODO: Fill this model card
+"""
+
+
+def _create_model_card(repo_dir: Path):
+    """Creates a model card for the repository.
+
+    TODO: Add metrics to model-index
+    TODO: Use information from common model cards
+    """
+    readme_path = repo_dir / "README.md"
+    prev_readme = ""
+    if readme_path.exists():
+        with readme_path.open("r", encoding="utf8") as f:
+            prev_readme = f.read()
+    with readme_path.open("w", encoding="utf-8") as f:
+        f.write(README_TEMPLATE)
+        f.write(prev_readme)
+
+
+_ALLOWLIST_PATHS = ["vocabulary", "config.json", "weights.th", "best.th", "metrics.json", "log"]
+
+
+def _copy_allowed_file(filepath: Path, dst_directory: Path):
+    """
+    Copies files from allowlist to a directory, overriding existing
+    files or directories if any.
+    """
+    if filepath.name not in _ALLOWLIST_PATHS:
+        return
+
+    dst = dst_directory / filepath.name
+    if dst.is_dir():
+        shutil.rmtree(str(dst))
+    elif dst.is_file():
+        dst.unlink()
+    if filepath.is_dir():
+        shutil.copytree(filepath, dst)
+    elif filepath.is_file():
+        if filepath.name == "best.th":
+            dst = dst_directory / "model.th"
+        shutil.copy(str(filepath), str(dst))
+
+
+def push_to_hf(
+    archive_path: Union[str, Path],
+    repo_name: str,
+    organization: Optional[str] = None,
+    commit_message: str = "Update repository",
+    local_repo_path: Union[str, Path] = "hub",
```

Comment:
I would expect a lot of the time, people don't care about having a local copy of the repo. We need one to do the work, but it could be a temporary directory that gets deleted afterwards, right?

Reply:
Yes, temporary directory is actually the approach we took in `sentence-transformers`. In `spaCy` we did something similar to the current state of the PR.

The advantage of having a local repo saved is that you can easily add new files or modify the model card without having to do an additional clone or going to the interface. 

You can think of this as a similar workflow to GitHub development. You have your local clone of your Hub repos, you do the changes and then you push. The `push_to_hub` command will create the initial repo and upload the model, but I do imagine many people wanting to do further, more custom changes. WDYT?